elasticdl/doc/async_sgd_design.md

```diff
@@ -0,0 +1,115 @@
+2. There may be multiple `ReportGradient` running concurrently. Different model variables may apply these gradients in different orders. 
+
+Also, the concurrent updates to variables in `ReportGradient` may cause some gradients are not applied, as the updates can overwritten by other concurrent running updates. TensorFlow optimizers have an argument [`use_locking`](https://github.com/tensorflow/tensorflow/blob/ff441191277b7e758deb48e45249fee9e880f2c8/tensorflow/python/training/optimizer.py#L319). If [`use_locking`](https://github.com/tensorflow/tensorflow/blob/ff441191277b7e758deb48e45249fee9e880f2c8/tensorflow/python/training/optimizer.py#L319) is `True`, TensorFlow will use a [lock](https://github.com/tensorflow/tensorflow/blob/11e22c01eb801ff24200afcdce8a03a7cdd2ed3f/tensorflow/core/kernels/training_ops.cc#L528) to prevent concurrent updates to variables.
+
+### Staleness in Asynchronous SGD
+In `ReportGradient`, the argument `version` may be smaller `PS_model_version`. 
+Staleness value is the difference between `PS_model_version` and `version`:
+
+```
+staleness = PS_model_version - version
+```
+
+According to some [researches](https://arxiv.org/abs/1810.03264), this staleness affects the training convergence, and large staleness may result in poor training accuracy. The deeper the model, the more impact of the staleness. Some optimizers such as [SGD](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/optimizers/SGD) and [Adagrad](https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/optimizers/Adagrad) are more robust to staleness, some optimizers such as other with momentum are very bad with staleness.
+
+[Staleness-aware asychronous SGD](https://arxiv.org/abs/1511.05950) proposes a method to modulate learning rate by the staleness. If the staleness if not 0, this method modulates the learning rate used in the optimizer as:
+
+```
+if staleness > 0:
+    learning_rate_used = learning_rate / staleness
+else:
+    learning_rate_used = learning_rate
+```
+
+### Stale Synchronous Parallel (SSP)
+In the pesudocode for the asynchronous SGD worker, the worker pulls model from PS in every minibatch step. [Stale synchronous parallel (SSP) method](https://dl.acm.org/citation.cfm?id=2999748) uses the strategy that the fastest worker can exceed the slowest one within a predefined staleness threshold. SSP can reduce the number of `get_model_from_ps` calls. The worker training process is:
+
+```
+staleness_threshold = predefined_staleness_threshold
+local_model, model_version = get_model_from_ps()
+local_update_count = 0
+for minibatch in training_data:
+    gradients = compute_gradient(local_model, minibatch)
+    apply_gradient(local_model, gradients)
+    report_gradient_to_ps(gradients, model_version)
+    local_update_count += 1
+    if local_update_count >= staleness_threshold:
```

Comment:
It seems the logic here is not correct. 
Let's image a corner case, one server and two workers. worker0 is very slow, and work1 is very fast.
At first, model_version = 0. Then work1 begin to run, it keeps updating parameters. Work0 is so slow that it does not send any gradients to server.
So, the `local_update_count` in work1 is the same with `model_version`. When `local_update_count>= staleness_threshold`, it will reset again, and keep working.
At some point, work0 finishes its first computation, it will send gradient to server, and the server will accept it! This is not expected.

The logic should be


Worker side
```
local_model, local_version = get_model_from_ps()
for minibatch in training_data:
    current_version = get_version_from_ps()
    if current_version - local_version > staleness_threshold:
        local_model, local_version = get_model_from_ps()
    gradients = compute_gradient(local_model, minibatch)
    report_gradient_to_ps(gradients)
```

Server side:

```
while true:
    gradients = get_gradients_from_queue()
    apply_gradients(model, gradients)
    model_version++;
```

Reply:
The logic should be following:

There is a shared global clock table, each work has its own clock count. We init the clock table to zero first.
```
init_clock_table()
```

Then, for each worker:


```
for minibatch in training_data:
    min_clock = get_min_clock_from_table()
    local_clock = get_clock_from_table(worker_id)
    if (local_clock - min_clock < staleness_threshold):
        model = get_model_ps()
        gradients = compute_gradient(model, minibatch)
        report_gradient_to_ps(gradients)
        inc_clock_table(worker_id)
```

Reply:
Parameter server should reject gradients from workers which lags behind it for a long time.

Reply:
Yes, in synchronous SGD, PS will reject gradients with a very old version. 
In asynchronous SGD, PS will accept all gradients. If staleness aware learning rate modulation is on, PS will modulate the learning rate with the staleness.

Reply:
> The logic should be following:
> 
> There is a shared global clock table, each work has its own clock count. We init the clock table to zero first.
> 
> ```
> init_clock_table()
> ```
> 
> Then, for each worker:
> 
> ```
> for minibatch in training_data:
>     min_clock = get_min_clock_from_table()
>     local_clock = get_clock_from_table(worker_id)
>     if (local_clock - min_clock < staleness_threshold):
>         model = get_model_ps()
>         gradients = compute_gradient(model, minibatch)
>         report_gradient_to_ps(gradients)
>         inc_clock_table(worker_id)
> ```

This is ESSP [(Eager Staleness Synchronous Parallel)](https://arxiv.org/pdf/1410.8043.pdf), which requires communication among workers (or indirectly through master or global storage). ESSP will reduce the staleness compared to SSP, but in the price of the worker clock communication. This clock communication will be very complexed as ElasticDL can add or remover workers during training. So we will not add ESSP feature in the first version of asynchronous SGD. 

ESSP/SSP is usually used in synchronous SGD. We can also use this to reduce PS/worker communication in asynchronous SGD. There are no experiments to show the effectiveness of  ESSP/SSP + async SGD training yet.